p2p/libp2p/netMessenger.go

```diff
@@ -613,32 +629,55 @@ func (netMes *networkMessenger) SetPeerShardResolver(peerShardResolver p2p.PeerS
 		return p2p.ErrNilPeerShardResolver
 	}
 
-	kadSharder, err := networksharding.NewKadSharder(prioBits, peerShardResolver)
+	reconnecter, _ := netMes.peerDiscoverer.(p2p.Reconnecter)
+	sharderFactory := factory.NewSharderFactory(
+		reconnecter,
+		peerShardResolver,
+		prioBits,
+		peer.ID(netMes.ID()),
+		netMes.targetConnCount,
+	)
+
+	kadSharder, err := sharderFactory.Create()
 	if err != nil {
 		return err
 	}
 
-	return netMes.connMonitor.SetSharder(kadSharder)
+	err = netMes.connMonitor.SetSharder(kadSharder)
+	if err != nil {
+		return err
+	}
+	netMes.peerShardResolver = peerShardResolver
+
+	return nil
 }
 
 // GetConnectedPeersInfo gets the current connected peers information
 func (netMes *networkMessenger) GetConnectedPeersInfo() *p2p.ConnectedPeersInfo {
+	if netMes.peerShardResolver == nil {
+		return &p2p.ConnectedPeersInfo{
+			UnknownPeers:    make([]string, 0),
+			IntraShardPeers: make([]string, 0),
+			CrossShardPeers: make([]string, 0),
+		}
+	}
+
 	peers := netMes.ctxProvider.connHost.Network().Peers()
 	peerInfo := &p2p.ConnectedPeersInfo{}
-	crt := netMes.connMonitor.sharder.GetShard(netMes.ctxProvider.connHost.ID())
+	selfShardId := netMes.peerShardResolver.ByID(netMes.ID())
```

Comment:
you could use an unknownPeerShardResolver implementation as constructur, than if is not needed in the first line of this method.

Reply:
Good point, remove the nil condition testing from all over the place